File: apps/wear-watchface/app/src/test/java/com/islamicdaydial/watchface/core/RingTest.kt
package com.islamicdaydial.watchface.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Date

class RingTest {

    private val lastMaghrib = Date(1742052600000L) // 2025-03-15T15:30:00.000Z
    private val nextMaghrib = Date(1742139000000L) // 2025-03-16T15:30:00.000Z

    @Test
    fun getIslamicDayProgress_returns0AtLastMaghrib() {
        assertEquals(0.0, getIslamicDayProgress(lastMaghrib, lastMaghrib, nextMaghrib), 1e-10)
    }

    @Test
    fun getIslamicDayProgress_returns1AtNextMaghrib() {
        assertEquals(1.0, getIslamicDayProgress(nextMaghrib, lastMaghrib, nextMaghrib), 1e-10)
    }

    @Test
    fun getIslamicDayProgress_returns05AtMidpoint() {
        val mid = Date((lastMaghrib.time + nextMaghrib.time) / 2)
        assertEquals(0.5, getIslamicDayProgress(mid, lastMaghrib, nextMaghrib), 1e-10)
    }

    @Test
    fun getIslamicDayProgress_clampsTo0ForTimesBeforeLastMaghrib() {
        val before = Date(lastMaghrib.time - 60000)
        assertEquals(0.0, getIslamicDayProgress(before, lastMaghrib, nextMaghrib), 1e-10)
    }

    @Test
    fun getIslamicDayProgress_clampsTo1ForTimesAfterNextMaghrib() {
        val after = Date(nextMaghrib.time + 60000)
        assertEquals(1.0, getIslamicDayProgress(after, lastMaghrib, nextMaghrib), 1e-10)
    }

    @Test
    fun timestampToAngle_returns0AtLastMaghrib() {
        assertEquals(0.0, timestampToAngle(lastMaghrib, lastMaghrib, nextMaghrib), 1e-5)
    }

    @Test
    fun timestampToAngle_returns360AtNextMaghrib() {
        assertEquals(360.0, timestampToAngle(nextMaghrib, lastMaghrib, nextMaghrib), 1e-5)
    }

    @Test
    fun timestampToAngle_returns180AtMidpoint() {
        val mid = Date((lastMaghrib.time + nextMaghrib.time) / 2)
        assertEquals(180.0, timestampToAngle(mid, lastMaghrib, nextMaghrib), 1e-5)
    }

    @Test
    fun getMarkers_returns7Markers() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val markers = getMarkers(timeline)
        assertEquals(7, markers.size)
    }

    @Test
    fun getMarkers_firstMarkerIsMaghribAt0() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val markers = getMarkers(timeline)
        assertEquals("maghrib", markers[0].id)
        assertEquals(0.0, markers[0].angleDeg, 1e-5)
        assertEquals("primary", markers[0].kind)
    }

    @Test
    fun getMarkers_markersInAscendingAngleOrder() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val markers = getMarkers(timeline)
        for (i in 1 until markers.size) {
            assertTrue(markers[i].angleDeg > markers[i - 1].angleDeg)
        }
    }

    @Test
    fun getMarkers_secondaryMarkerIsLastThirdStart() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val markers = getMarkers(timeline)
        val secondary = markers.filter { it.kind == "secondary" }
        assertEquals(listOf("last_third_start"), secondary.map { it.id })
    }

    @Test
    fun getRingSegments_returns8Segments() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val segments = getRingSegments(timeline)
        assertEquals(8, segments.size)
    }

    @Test
    fun getRingSegments_firstSegmentStartsAt0() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val segments = getRingSegments(timeline)
        assertEquals(0.0, segments[0].startAngleDeg, 1e-5)
    }

    @Test
    fun getRingSegments_lastSegmentEndsAt360() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val segments = getRingSegments(timeline)
        assertEquals(360.0, segments[segments.size - 1].endAngleDeg, 1e-5)
    }

    @Test
    fun getRingSegments_segmentsAreContiguous() {
        val timeline = ComputedTimeline(
            lastMaghrib = lastMaghrib,
            isha = Date(1742058000000L),
            islamicMidnight = Date(1742069100000L),
            lastThirdStart = Date(1742074200000L),
            fajr = Date(1742090400000L),
            sunrise = Date(1742094900000L),
            dhuhr = Date(1742116500000L),
            asr = Date(1742127000000L),
            nextMaghrib = nextMaghrib
        )
        val segments = getRingSegments(timeline)
        for (i in 1 until segments.size) {
            assertEquals(
                segments[i - 1].endAngleDeg,
                segments[i].startAngleDeg,
                1e-5
            )
        }
    }
}

File: packages/core/src/__tests__/ring.test.ts
import { describe, it, expect } from 'vitest';
import { getIslamicDayProgress, timestampToAngle, getMarkers, getRingSegments } from '../ring.js';
import type { ComputedTimeline } from '../types.js';

const lastMaghrib = new Date('2025-03-15T15:30:00.000Z');
const nextMaghrib = new Date('2025-03-16T15:30:00.000Z');

describe('getIslamicDayProgress', () => {
  it('returns 0 at lastMaghrib', () => {
    expect(getIslamicDayProgress(lastMaghrib, lastMaghrib, nextMaghrib)).toBe(0);
  });

  it('returns 1 at nextMaghrib', () => {
    expect(getIslamicDayProgress(nextMaghrib, lastMaghrib, nextMaghrib)).toBe(1);
  });

  it('returns 0.5 at midpoint', () => {
    const mid = new Date((lastMaghrib.getTime() + nextMaghrib.getTime()) / 2);
    expect(getIslamicDayProgress(mid, lastMaghrib, nextMaghrib)).toBeCloseTo(0.5, 10);
  });

  it('clamps to 0 for times before lastMaghrib', () => {
    const before = new Date(lastMaghrib.getTime() - 60000);
    expect(getIslamicDayProgress(before, lastMaghrib, nextMaghrib)).toBe(0);
  });

  it('clamps to 1 for times after nextMaghrib', () => {
    const after = new Date(nextMaghrib.getTime() + 60000);
    expect(getIslamicDayProgress(after, lastMaghrib, nextMaghrib)).toBe(1);
  });
});

describe('timestampToAngle', () => {
  it('returns 0 at lastMaghrib', () => {
    expect(timestampToAngle(lastMaghrib, lastMaghrib, nextMaghrib)).toBe(0);
  });

  it('returns 360 at nextMaghrib', () => {
    expect(timestampToAngle(nextMaghrib, lastMaghrib, nextMaghrib)).toBe(360);
  });

  it('returns 180 at midpoint', () => {
    const mid = new Date((lastMaghrib.getTime() + nextMaghrib.getTime()) / 2);
    expect(timestampToAngle(mid, lastMaghrib, nextMaghrib)).toBeCloseTo(180, 5);
  });
});

describe('getMarkers', () => {
  const timeline: ComputedTimeline = {
    lastMaghrib: new Date('2025-03-15T15:30:00.000Z'),
    isha: new Date('2025-03-15T17:00:00.000Z'),
    lastThirdStart: new Date('2025-03-15T22:30:00.000Z'),
    fajr: new Date('2025-03-16T02:00:00.000Z'),
    sunrise: new Date('2025-03-16T03:15:00.000Z'),
    duhaStart: new Date('2025-03-16T03:35:00.000Z'),
    duhaEnd: new Date('2025-03-16T09:10:00.000Z'),
    dhuhr: new Date('2025-03-16T09:15:00.000Z'),
    asr: new Date('2025-03-16T12:30:00.000Z'),
    nextMaghrib: new Date('2025-03-16T15:30:00.000Z'),
  };

  it('returns 9 markers', () => {
    const markers = getMarkers(timeline);
    expect(markers).toHaveLength(9);
  });

  it('first marker is maghrib at 0°', () => {
    const markers = getMarkers(timeline);
    expect(markers[0].id).toBe('maghrib');
    expect(markers[0].angleDeg).toBe(0);
    expect(markers[0].kind).toBe('primary');
  });

  it('markers are in ascending angle order', () => {
    const markers = getMarkers(timeline);
    for (let i = 1; i < markers.length; i++) {
      expect(markers[i].angleDeg).toBeGreaterThan(markers[i - 1].angleDeg);
    }
  });

  it('secondary markers are last_third_start, duha_start, duha_end', () => {
    const markers = getMarkers(timeline);
    const secondary = markers.filter(m => m.kind === 'secondary');
    expect(secondary.map(m => m.id)).toEqual(['last_third_start', 'duha_start', 'duha_end']);
  });
});

describe('getRingSegments', () => {
  const timeline: ComputedTimeline = {
    lastMaghrib: new Date('2025-03-15T15:30:00.000Z'),
    isha: new Date('2025-03-15T17:00:00.000Z'),
    lastThirdStart: new Date('2025-03-15T22:30:00.000Z'),
    fajr: new Date('2025-03-16T02:00:00.000Z'),
    sunrise: new Date('2025-03-16T03:15:00.000Z'),
    duhaStart: new Date('2025-03-16T03:35:00.000Z'),
    duhaEnd: new Date('2025-03-16T09:10:00.000Z'),
    dhuhr: new Date('2025-03-16T09:15:00.000Z'),
    asr: new Date('2025-03-16T12:30:00.000Z'),
    nextMaghrib: new Date('2025-03-16T15:30:00.000Z'),
  };

  it('returns 7 segments', () => {
    const segments = getRingSegments(timeline);
    expect(segments).toHaveLength(7);
  });

  it('first segment starts at 0°', () => {
    const segments = getRingSegments(timeline);
    expect(segments[0].startAngleDeg).toBe(0);
  });

  it('last segment ends at 360°', () => {
    const segments = getRingSegments(timeline);
    expect(segments[segments.length - 1].endAngleDeg).toBeCloseTo(360, 5);
  });

  it('segments are contiguous (no gaps)', () => {
    const segments = getRingSegments(timeline);
    for (let i = 1; i < segments.length; i++) {
      expect(segments[i].startAngleDeg).toBeCloseTo(segments[i - 1].endAngleDeg, 5);
    }
  });
});
